### DISCUSSION TITLE
Unit Testing with Semantic Kernel

### DISCUSSION THREAD

**Discussion by Manish-22** [microsoft/semantic-kernel#8088] [Q&A] [answered]
Hi team,
I have been looking at this document for testing my code.
https://devblogs.microsoft.com/semantic-kernel/unit-testing-with-semantic-kernel/

```
[Fact] 
public async Task DoWorkWithPrompt() 
{ 
    // Arrange 
    var mockChatCompletion = new Mock<IChatCompletionService>(); 
    mockChatCompletion 
        .Setup(x => x.GetChatMessageContentsAsync( 
            It.IsAny<ChatHistory>(), 
            It.IsAny<PromptExecutionSettings>(), 
            It.IsAny<Kernel>(), 
            It.IsAny<CancellationToken>())) 
        .ReturnsAsync([new ChatMessageContent(AuthorRole.Assistant, "AI response")]); 
 
    var kernelBuilder = Kernel.CreateBuilder(); 
    kernelBuilder.Services.AddSingleton(mockChatCompletion.Object); 
 
    var kernel = kernelBuilder.Build(); 
    var service = new MyService(kernel); 
 
    // Act 
    var result = await service.DoWorkAsync("Prompt to AI"); 
 
    // Assert 
    Assert.Equal("AI response", result.ToString()); 
} 
````

Here the AI response is a simple string and that code path is covered by the test.
What if there is a funtion call by the AI. How can i mock it.
I'm assuming i need to pass something over here [new ChatMessageContent(AuthorRole.Assistant, "AI response")]

Could you let me know how can i pass the funtion to be called in the mock?

**Comment by markwallace-microsoft** [comment, marked as answer]
Hi @Manish-22 

you can mock at the HHTP level to control the response. 

Here is an example (for Mistral AI): https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs#L168

Here are the sample responses: https://github.com/microsoft/semantic-kernel/tree/main/dotnet/src/Connectors/Connectors.MistralAI.UnitTests/TestData

**Reply by Manish-22** [reply]
Thanks @markwallace-microsoft
